File: opengl/tpixmap.java

class TPixmapJava {
	static public void DataToPixmap( BBDataBuffer imagedata, BBDataBuffer buf, int[] info ) {
		Bitmap bitmap=null;
		try{
			//byte[] bytes = new byte[imagedata.GetByteBuffer().capacity()];
			//imagedata.GetByteBuffer().get(bytes, 0, bytes.length);
			//bitmap=BitmapFactory.decodeByteArray( bytes,0, bytes.length);
			bitmap=BitmapFactory.decodeByteArray( imagedata.GetByteBuffer().array(),0, imagedata.GetByteBuffer().array().length);
		}catch( OutOfMemoryError e ){
			throw new Error( "Out of memory error loading bitmap" );
		}
		if( bitmap==null ) return;
		
		int width=bitmap.getWidth(),height=bitmap.getHeight();

		int size=width*height;
		int[] pixels=new int[size];
		bitmap.getPixels( pixels,0,width,0,0,width,height );
		
		buf.Discard();
		boolean result = buf._New( size*4 );
		if (!result) throw new Error( "Out of memory error loading bitmap" );
		
		for( int i=0;i<size;++i ){
			int p=pixels[i];
			int a=(p>>24) & 255;
			int r=(p>>16) & 255;
			int g=(p>>8) & 255;
			int b=p & 255;
			buf.PokeInt( (i<<2),(a<<24)|(b<<16)|(g<<8)|r ); //i*4
		}
		
		if( info.length>0 ) info[0]=width;
		if( info.length>1 ) info[1]=height;
	}
}

File: opengl/tpixmap.cpp


void DataToPixmap( BBDataBuffer* imagedata, BBDataBuffer* buf, Array<int> info ) {
	
	int width,height,depth;
	int len = imagedata->Length();
	unsigned char *data=loadImage( (unsigned char*)(imagedata->ReadPointer(0)) ,len, &width,&height,&depth );
	
//printf("xx %i %i %i , %i\n",width,height, depth, len); fflush(stdout);
	if( !data || depth<1 || depth>4 ) return;


	
	int size=width*height;
	buf->Discard();
	//buf= new BBDataBuffer;
	buf->_New( size*4 );
	
	unsigned char *src=data,*dst=(unsigned char*)buf->WritePointer(0);
	int i;
	
	//convert to ARGB for all depths
	switch( depth ){
	case 1:for( i=0;i<size;++i ){ *dst++=*src;*dst++=*src;*dst++=*src++;*dst++=255; } break;
	case 2:for( i=0;i<size;++i ){ *dst++=*src;*dst++=*src;*dst++=*src++;*dst++=*src++; } break;
	case 3:for( i=0;i<size;++i ){ *dst++=*src++;*dst++=*src++;*dst++=*src++;*dst++=255; } break;
	case 4:for( i=0;i<size;++i ){ *dst++=*src++;*dst++=*src++;*dst++=*src++;*dst++=*src++; } break;
	}
	
	if( info.Length()>0 ) info[0]=width;
	if( info.Length()>1 ) info[1]=height;
	
	unloadImage( data );
	
	//return buf;

}

/*
void LoadImageData( BBDataBuffer *buf, String path,Array<int> info ){
	int width,height,depth;
	unsigned char *data=loadImage( path,&width,&height,&depth );
	if( !data || depth<1 || depth>4 ) return;
	
	int size=width*height;
	buf->Discard();
	//buf= new BBDataBuffer;
	buf->_New( size*4 );
	
	unsigned char *src=data,*dst=(unsigned char*)buf->WritePointer();
	int i;
	
	switch( depth ){
	case 1:for( i=0;i<size;++i ){ *dst++=*src;*dst++=*src;*dst++=*src++;*dst++=255; } break;
	case 2:for( i=0;i<size;++i ){ *dst++=*src;*dst++=*src;*dst++=*src++;*dst++=*src++; } break;
	case 3:for( i=0;i<size;++i ){ *dst++=*src++;*dst++=*src++;*dst++=*src++;*dst++=255; } break;
	case 4:for( i=0;i<size;++i ){ *dst++=*src++;*dst++=*src++;*dst++=*src++;*dst++=*src++; } break;
	}
	
	if( info.Length()>0 ) info[0]=width;
	if( info.Length()>1 ) info[1]=height;
	
	unloadImage( data );
	
	//return buf;
}
*/